Archive: Add dynamic paths to browser tab

956MB opened the pull request:
# What's new

Adds to the option to show current directory/path in archive tab instead of static tab text. Default state is `OFF` to maintain the existing look, but I personally think `Current` could probably be set as default if merged. Of course all text involed `Show Browser Path`, `Current`, `Brief` and `Full` could be changed as well.

- Adds setting under "Interface > File Browser > Show Browser Path" with four options:
  - Off (default): Shows static tab text like before (e.g. "Browser", "Sub-GHz", etc.)
  - Current: Shows the current directory name only (e.g. "Games")
  - Brief: Abbreviates the full path to the current directory (e.g. /e/a/Games")
  - Full: Shows complete path from root (e.g. "/ext/apps/Games")
- Path updates when navigating through directories. It is also maintained when opening an app and clicking back into the browser.
- Maintains existing header styling and layout keeping it within the frame
- Long paths that exceed the frame width are drawn with `elements_scrollable_text_line` to allow for full visibility.
    - e.g. Short paths like `Games` are drawn centered like normal, and long ones like `/ext/apps/Games` scroll.
    
## Screenshots

#### "Off"

<img width="528" alt="OFF" src="https://github.com/user-attachments/assets/4951fa20-2335-48a0-9a72-cde5a817f163" />

#### "Current"

<img width="528" alt="Current" src="https://github.com/user-attachments/assets/d6bac53f-3d04-4312-8b2f-f2cb80245a83" />

#### "Brief"

![Brief](https://github.com/user-attachments/assets/649aa595-3d22-4792-b9e3-83624d734f41)

#### "Full"

![Full](https://github.com/user-attachments/assets/cef5eeee-cf02-4d2a-815c-056ce90d6637)

-----
# For the reviewer

- [x] I've uploaded the firmware with this patch to a device and verified its functionality
- [x] I've confirmed the bug to be fixed / feature to be stable

956MB committed:
feat(Archive): Adds dynamic paths to browser tab

I thought it would be cool to be able to see when I was in say the
`apps_data/flipchess` folder, or some other nested folder, and have
"flipchess" be shown rather than the just "Browser" all the time. It's
not a huge or really crucial change, just a personal modification that
some people may also find useful.

WillyJL commented:
*missed opportunity for `/e/a/Sports` in the brief example xD*

sorry i just had to :laughing:, i like this idea! will review in the coming days

956MB commented:
> _missed opportunity for `/e/a/Sports` in the brief example xD_

AGH! Scrap the whole thing. Kidding, looking forward to it 👍

956MB committed:
fix(Archive): Clean up, built-ins, safer

Cleaned up `archive_update_formatted_path` function using switch case,
and also using the built in furi functions where possible. Also removed
the arbitrary and unnecessary `path_buf[256]`, making formatting the
path in the Brief section more memory safe.

WillyJL committed:
Merge remote-tracking branch 'mntm/dev' into mntm-pr-322

WillyJL commented on lib/momentum/settings.c:
```diff
@@ -100,6 +101,7 @@ static const struct {
     {setting_bool(sort_dirs_first)},
     {setting_bool(show_hidden_files)},
     {setting_bool(show_internal_tab)},
+    {setting_bool(show_browser_path)},
```
using `setting_bool()` here will enforce the value is either 0 or 1, and save to file as "true" or "false". should be `setting_enum()`, which also specifies the max value (discards values that arent 0 to max-1). also since its a multiple option setting, could be renamed to `browser_path_mode` to match the enum.

WillyJL committed:
Rename setting and use setting_enum()

956MB commented on applications/main/archive/views/archive_browser_view.c:
```diff
@@ -293,7 +347,20 @@ static void archive_render_status_bar(Canvas* canvas, ArchiveBrowserViewModel* m
     canvas_draw_line(canvas, 1, 11, 49, 11); // shadow bottom
-    canvas_draw_str_aligned(canvas, 25, 9, AlignCenter, AlignBottom, tab_name);
+
+    size_t text_width = canvas_string_width(canvas, tab_name);
```
I just realized looking at this that checking the width and using canvas_draw_str_aligned isnt necessary because scrollable_text_line will function the same when it doesn't need to scroll.

WillyJL replied:
yep :D

WillyJL replied:
actually no, if its shorter, then the scrollable text would align it to left instead of centered. however, there is a centered variant of the scrollable text function

WillyJL commented on applications/main/archive/views/archive_browser_view.c:
```diff
@@ -58,6 +58,57 @@ void archive_browser_set_callback(
+    }
+
+    ArchiveTabEnum tab = archive_get_tab(browser);
+    if(!momentum_settings.show_browser_path || archive_is_home(browser)) {
```
here value 0 coincides with off, but being an enum (and also considering renaming the setting as before) its better to explicitly check against enum value

WillyJL commented on applications/main/archive/views/archive_browser_view.c:
```diff
@@ -58,6 +58,57 @@ void archive_browser_set_callback(
+            break;
+        }
+
+        default:
```
better to specify the enum value here too, in case there are more values added later, also makes it clearer

WillyJL commented on applications/main/archive/views/archive_browser_view.c:
```diff
@@ -58,6 +58,57 @@ void archive_browser_set_callback(
+        }
+    }
+
+    furi_string_set(browser->prev_path, browser->path);
```
this doesn't seem to be used anywhere, lost between refactoring? ;)

WillyJL reviewed (changes requested):
small things, working on them rn

WillyJL committed:
Merge remote-tracking branch 'mntm/dev' into mntm-pr-322

WillyJL committed:
Remove prev_path

WillyJL committed:
Enum names

WillyJL committed:
Single text draw call

WillyJL committed:
Rename to statusbar title

WillyJL commented:
hmm tho the way i did it makes it not change text while searching. just that the scrollable text text functions need a furistring so i thought to just use the same one as formatted path, but then for searching it needs to change even if the path didnt change. might be easier to just use a temporary furistring, or to use the normal draw function while searching

WillyJL committed:
Revert "Rename to statusbar title"

This reverts commit 243f2a754ff19e4d45eb7d287500cd19739bb8b6.

WillyJL committed:
Revert "Single text draw call"

This reverts commit 3bbe30449ba9566ca779c685b63fc452ba630dc8.

WillyJL committed:
Simpler fix

WillyJL committed:
Shorter setting label

WillyJL reviewed (approved):
nice :D

WillyJL committed:
Update changelog
